hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -186,6 +186,34 @@ object LowerTableIR {
           val env: Env[IR] = Env("row" -> row, "global" -> loweredChild.globals)
           loweredChild.copy(body = StreamMap(loweredChild.body, row.name, Subst(newRow, BindingEnv(env, scan = Some(env)))))
 
+        case t@TableKeyBy(child, newKey, isSorted: Boolean) =>
+          val loweredChild = lower(child)
+          val nPreservedFields = loweredChild.partitioner.kType.fieldNames
+            .zip(newKey)
+            .takeWhile { case (l, r) => l == r }
+            .length
+
+          require(!isSorted || nPreservedFields > 0 || newKey.isEmpty)
+
+          if (nPreservedFields == newKey.length)
+            loweredChild
+          else if (isSorted) {
+            val newPartitioner = loweredChild.partitioner
+              .coarsen(nPreservedFields)
+              .extendKey(t.typ.keyType)
+            if (newPartitioner.isStrict)
+              loweredChild.copy(partitioner = newPartitioner)
+            else throw new LowererUnsupportedOperation("TableKeyBy with isSorted but a non-strict partitioner requires partitioner adjustments")
```

Comment:
I'm confused why this is necessary.  I thought partitioners allowed the endpoints to overlap (and some downstream consumer would have to sort it out).

Reply:
I don't think so. We could redesign things in TableStage.